### PR TITLE
Search: Investigate frontend search options (again)

### PR DESCRIPTION
**Pull request by ryantxu**
When we started the search project, one possible solution was trying a pure frontend search implementation.  We tried that with https://github.com/lucaong/minisearch, and it worked OK, but the initial startup and complexity did not seem worth it, and we were trying more complete document search (with description and complex scoring)

Now that our interfaces have stabilized a bit, and we can more easily test/compare multiple implementations, I think it is worth considering again.  

This PR does not add anything too interesting, other than a hack/flag to kick off simple substring query.

Can test with:
http://localhost:3000/?orgId=1&search=open&query=heat&do-frontend-query

@leeoniya aims to try updating this with a more robust fuzzy search (I think in a new PR)

### CODE REVIEW

File: public/app/features/search/service/frontend.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  full = new Map<string, FullResultCache>();
+
+  constructor(private parent: GrafanaSearcher) {}
+
+  async search(query: SearchQuery): Promise<QueryResponse> {
+    if (query.facet?.length) {
+      throw new Error('facets not supported!');
+    }
+    // Don't bother... not needed for this exercise
+    if (query.tags?.length || query.ds_uid?.length) {
```
Comment (ArturWierzbicki) [inline public/app/features/search/service/frontend.ts:16]: or if panel search?

File: public/app/features/search/service/frontend.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    };
+  }
+
+  async getDashboards(): Promise<FullResultCache> {
+    let res = this.full.get('dashboard');
+    if (res) {
+      return Promise.resolve(res);
+    }
+    const v = await this.parent.search({
+      kind: ['dashboard'],
```
Comment (ArturWierzbicki) [inline public/app/features/search/service/frontend.ts:37]: shouldn't we include folders in the cache?
`      kind: ['dashboard', 'folder'],`

File: public/app/features/search/service/frontend.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  async tags(query: SearchQuery): Promise<TermCount[]> {
+    return this.parent.tags(query);
+  }
+}
+
+class FullResultCache {
+  readonly lower: string[];
+  empty: DataFrameView<DashboardQueryResult>;
+
+  constructor(private full: DataFrameView<DashboardQueryResult>) {
```
Comment (ArturWierzbicki) [inline public/app/features/search/service/frontend.ts:64]: should we sort the dashboards alphabetically or is that part of @leeoniya's PR?

File: public/app/features/search/service/frontend.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
+import { DataFrameView, SelectableValue, ArrayVector } from '@grafana/data';
+import { TermCount } from 'app/core/components/TagFilter/TagFilter';
+
+import { DashboardQueryResult, GrafanaSearcher, QueryResponse, SearchQuery } from '.';
+
+export class FrontendSearcher implements GrafanaSearcher {
```
Comment (ArturWierzbicki) [inline public/app/features/search/service/frontend.ts:6]: im getting a weird result for empty search:
image


the `*` is probably added somewhere as a fallback

**Review by ArturWierzbicki — APPROVED** [review]
experiments!!

File: public/app/features/search/service/frontend.ts (hunk 3 above)
Comment (leeoniya) [reply public/app/features/search/service/frontend.ts:64]: the ranking pass will will be able to sort them alphabetically, so there's no need to do it in a prep step.

do we have any other attributes that should affect sorting? e.g. created, last modified, recently used, etc.?

File: public/app/features/search/service/frontend.ts (hunk 4 above)
Comment (ryantxu) [reply public/app/features/search/service/frontend.ts:6]: dooh -- yes

File: public/app/features/search/service/frontend.ts (hunk 3 above)
Comment (ryantxu) [reply public/app/features/search/service/frontend.ts:64]: maybe tags?  the rest we will want to use for direct sorting, but not in this first round

File: public/app/features/search/service/frontend.ts (hunk 1 above)
Comment (ryantxu) [reply public/app/features/search/service/frontend.ts:16]: I updated this so our cache matches the input request 👍

Comment (ryantxu) [reply public/app/features/search/service/frontend.ts:16]: so now it *will* do panel search

**Comment by sakjur** [comment]
I'm updating the milestone to v9.2.0 as this was merged before the v9.2.x branchoff happened.

**Pull request merged**